Compile error in opencl_core.cpp

I'm getting the following error in ```opencl_core.cpp``` when trying to compile OpenCV:

```
[ 10%] Building CXX object modules/core/CMakeFiles/opencv_core.dir/src/opencl/runtime/opencl_core.cpp.o
/home/user/abs/opencv-noffmpeg-git/src/opencv-noffmpeg-git/modules/core/src/opencl/runtime/opencl_core.cpp:210:9: error: ‘opencl_fn4’ was not declared in this scope
         opencl_fn4<OPENCL_FN_clSVMAlloc, void*, cl_context, cl_svm_mem_flags, size_t, unsigned int>::switch_fn;
         ^~~~~~~~~~
/home/user/abs/opencv-noffmpeg-git/src/opencv-noffmpeg-git/modules/core/src/opencl/runtime/opencl_core.cpp:213:9: error: ‘opencl_fn2’ was not declared in this scope
         opencl_fn2<OPENCL_FN_clSVMFree, void, cl_context, void*>::switch_fn;
         ^~~~~~~~~~
/home/user/abs/opencv-noffmpeg-git/src/opencv-noffmpeg-git/modules/core/src/opencl/runtime/opencl_core.cpp:216:9: error: ‘opencl_fn3’ was not declared in this scope
         opencl_fn3<OPENCL_FN_clSetKernelArgSVMPointer, cl_int, cl_kernel, cl_uint, const void*>::switch_fn;
         ^~~~~~~~~~
/home/user/abs/opencv-noffmpeg-git/src/opencv-noffmpeg-git/modules/core/src/opencl/runtime/opencl_core.cpp:225:9: error: ‘opencl_fn8’ was not declared in this scope
         opencl_fn8<OPENCL_FN_clEnqueueSVMMemcpy, cl_int, cl_command_queue, cl_bool, void*, const void*, size_t, cl_uint, const cl_event*, cl_event*>::switch_fn;
         ^~~~~~~~~~
/home/user/abs/opencv-noffmpeg-git/src/opencv-noffmpeg-git/modules/core/src/opencl/runtime/opencl_core.cpp:228:9: error: ‘opencl_fn8’ was not declared in this scope
         opencl_fn8<OPENCL_FN_clEnqueueSVMMemFill, cl_int, cl_command_queue, void*, const void*, size_t, size_t, cl_uint, const cl_event*, cl_event*>::switch_fn;
         ^~~~~~~~~~
/home/user/abs/opencv-noffmpeg-git/src/opencv-noffmpeg-git/modules/core/src/opencl/runtime/opencl_core.cpp:231:9: error: ‘opencl_fn8’ was not declared in this scope
         opencl_fn8<OPENCL_FN_clEnqueueSVMMap, cl_int, cl_command_queue, cl_bool, cl_map_flags, void*, size_t, cl_uint, const cl_event*, cl_event*>::switch_fn;
         ^~~~~~~~~~
/home/user/abs/opencv-noffmpeg-git/src/opencv-noffmpeg-git/modules/core/src/opencl/runtime/opencl_core.cpp:234:9: error: ‘opencl_fn5’ was not declared in this scope
         opencl_fn5<OPENCL_FN_clEnqueueSVMUnmap, cl_int, cl_command_queue, void*, cl_uint, const cl_event*, cl_event*>::switch_fn;
         ^~~~~~~~~~
make[2]: *** [modules/core/CMakeFiles/opencv_core.dir/build.make:883: modules/core/CMakeFiles/opencv_core.dir/src/opencl/runtime/opencl_core.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:988: modules/core/CMakeFiles/opencv_core.dir/all] Error 2
make: *** [Makefile:161: all] Error 2
```

This seems to be caused by some relatively recent code change because it was compiling well some days ago.

**System information (version):**
**OS:** Arch Linux x86_64
**OpenCV:** git master (currently at 3.1.0.r1475.g218f757)
**GCC:** 6.2.1
**CMake:** 3.6.2
**glibc:** 2.24
**opencl-headers:** 2.1.20151224
**opencl-nvidia:** 370.28
**CPU:** Intel Haswell
**GPU:** NVIDIA GTX 750 (using proprietary driver)

Commands used for building:
```
cmake \
	-DBUILD_SHARED_LIBS:BOOL=ON \
	\
	-DBUILD_opencv_apps:BOOL=ON             -DBUILD_opencv_calib3d:BOOL=ON  	-DBUILD_opencv_core:BOOL=ON \
	-DBUILD_opencv_features2d:BOOL=ON       -DBUILD_opencv_flann:BOOL=ON    	-DBUILD_opencv_highgui:BOOL=ON \
	-DBUILD_opencv_imgcodecs:BOOL=ON        -DBUILD_opencv_imgproc:BOOL=ON  	-DBUILD_opencv_ml:BOOL=ON \
	-DBUILD_opencv_objdetect:BOOL=ON        -DBUILD_opencv_photo:BOOL=ON    	-DBUILD_opencv_shape:BOOL=ON \
	-DBUILD_opencv_stitching:BOOL=ON        -DBUILD_opencv_superres:BOOL=ON 	-DBUILD_opencv_ts:BOOL=ON \
	-DBUILD_opencv_video:BOOL=ON            -DBUILD_opencv_videoio:BOOL=ON      -DBUILD_opencv_videostab:BOOL=ON \
	-DBUILD_opencv_world:BOOL=OFF \
	\
	-DBUILD_opencv_cudaarithm:BOOL=OFF      -DBUILD_opencv_cudabgsegm:BOOL=OFF      -DBUILD_opencv_cudacodec:BOOL=OFF \
	-DBUILD_opencv_cudafeatures2d:BOOL=OFF  -DBUILD_opencv_cudafilters:BOOL=OFF     -DBUILD_opencv_cudaimgproc:BOOL=OFF \
	-DBUILD_opencv_cudalegacy:BOOL=OFF      -DBUILD_opencv_cudaobjdetect:BOOL=OFF   -DBUILD_opencv_cudaoptflow:BOOL=OFF \
	-DBUILD_opencv_cudastereo:BOOL=OFF      -DBUILD_opencv_cudawarping:BOOL=OFF     -DBUILD_opencv_cudev:BOOL=OFF \
	\
	-DCUDA_64_BIT_DEVICE_CODE:BOOL=OFF \
	-DCUDA_FAST_MATH:BOOL=OFF \
	\
	-DCMAKE_BUILD_TYPE:STRING=Release \
	-DCMAKE_COLOR_MAKEFILE:BOOL=ON \
	-DCMAKE_INSTALL_PREFIX:PATH=/usr \
	-DCPACK_BINARY_DEB:BOOL=OFF \
	\
	-DPYTHON2_NUMPY_INCLUDE_DIRS:PATH=/usr/include/python2.7/numpy \
	-DPYTHON3_NUMPY_INCLUDE_DIRS:PATH=/usr/include/python3.5m/numpy \
	\
	-DENABLE_FAST_MATH:BOOL=ON \
	\
	-DENABLE_AVX:BOOL=OFF   -DENABLE_AVX2:BOOL=OFF   -DENABLE_FMA3:BOOL=OFF   -DENABLE_POPCNT:BOOL=OFF \
	-DENABLE_SSE:BOOL=ON    -DENABLE_SSE2:BOOL=ON    -DENABLE_SSE3:BOOL=ON    -DENABLE_SSSE3:BOOL=OFF \
    -DENABLE_SSE41:BOOL=OFF -DENABLE_SSE42:BOOL=OFF \
	\
	-DBUILD_EXAMPLES:BOOL=OFF \
	-DBUILD_PERF_TESTS:BOOL=OFF \
	-DBUILD_TESTS:BOOL=OFF \
	-DINSTALL_C_EXAMPLES:BOOL=OFF \
	-DINSTALL_PYTHON_EXAMPLES:BOOL=OFF \
	-DINSTALL_TESTS:BOOL=OFF \
	\
	-DWITH_1394:BOOL=ON       -DWITH_CLP:BOOL=OFF            -DWITH_CUBLAS:BOOL=OFF        -DWITH_CUDA:BOOL=OFF \
	-DWITH_CUFFT:BOOL=OFF     -DWITH_EIGEN:BOOL=ON           -DWITH_FFMPEG:BOOL=OFF        -DWITH_GDAL:BOOL=OFF \
	-DWITH_GIGEAPI:BOOL=OFF   -DWITH_GPHOTO2:BOOL=ON         -DWITH_GSTREAMER:BOOL=ON      -DWITH_GSTREAMER_0_10:BOOL=OFF \
	-DWITH_GTK:BOOL=ON        -DWITH_GTK_2_X:BOOL=ON         -DWITH_IPP_A:BOOL=ON          -DWITH_IPP:BOOL=ON \
	-DWITH_JASPER:BOOL=ON     -DWITH_JPEG:BOOL=ON            -DWITH_LIBV4L:BOOL=ON         -DWITH_NVCUVID:BOOL=ON \
	-DWITH_OPENCL:BOOL=ON     -DWITH_OPENCLAMDBLAS:BOOL=ON   -DWITH_OPENCLAMDFFT:BOOL=ON   -DWITH_OPENCL_SVM:BOOL=ON \
	-DWITH_OPENEXR:BOOL=ON    -DWITH_OPENGL:BOOL=ON          -DWITH_OPENMP:BOOL=ON         -DWITH_OPENNI:BOOL=OFF \
	-DWITH_OPENNI2:BOOL=OFF   -DWITH_PNG:BOOL=ON             -DWITH_PVAPI:BOOL=OFF         -DWITH_QT:BOOL=OFF \
	-DWITH_TBB:BOOL=ON        -DWITH_TIFF:BOOL=ON            -DWITH_UNICAP:BOOL=OFF        -DWITH_V4L:BOOL=ON \
	-DWITH_VTK:BOOL=OFF       -DWITH_WEBP:BOOL=ON            -DWITH_XIMEA:BOOL=OFF         -DWITH_XINE:BOOL=OFF \
	-DWITH_MATLAB:BOOL=OFF    -DWITH_OPENCLAMDFFT:BOOL=ON    -DWITH_PTHREADS_PF:BOOL=ON    -DWITH_VA_INTEL:BOOL=ON \
	\
	-DBUILD_JASPER:BOOL=OFF   -DBUILD_JPEG:BOOL=OFF          -DBUILD_OPENEXR:BOOL=OFF      -DBUILD_PNG:BOOL=OFF \
	-DBUILD_TIFF:BOOL=OFF     -DBUILD_ZLIB:BOOL=OFF \
	\
	-DOPENCV_WARNINGS_ARE_ERRORS:BOOL=OFF \
	-Wno-dev \
	.
	
	make -j1
```

Any help or comment would be appreciated.
Thank you.

--- a/modules/core/src/opencl/runtime/opencl_core.cpp
+++ b/modules/core/src/opencl/runtime/opencl_core.cpp
@@ -206,32 +206,33 @@ enum OPENCL_FN_SVM_ID
     OPENCL_FN_clEnqueueSVMUnmap,
 };
 
+opencl_fn4(OPENCL_FN_clSVMAlloc, void*, (cl_context p1, cl_svm_mem_flags p2, size_t p3, unsigned int p4))
 void* (CL_API_CALL *clSVMAlloc)(cl_context context, cl_svm_mem_flags flags, size_t size, unsigned int alignment) =
-        opencl_fn4<OPENCL_FN_clSVMAlloc, void*, cl_context, cl_svm_mem_flags, size_t, unsigned int>::switch_fn;
+        OPENCL_FN_clSVMAlloc_switch_fn;
 static const struct DynamicFnEntry _clSVMAlloc_definition = { "clSVMAlloc", (void**)&clSVMAlloc};
+opencl_fn2(OPENCL_FN_clSVMFree, void, (cl_context p1, void* p2))
 void (CL_API_CALL *clSVMFree)(cl_context context, void* svm_pointer) =
-        opencl_fn2<OPENCL_FN_clSVMFree, void, cl_context, void*>::switch_fn;
+        OPENCL_FN_clSVMFree_switch_fn;
 static const struct DynamicFnEntry _clSVMFree_definition = { "clSVMFree", (void**)&clSVMFree};
+opencl_fn3(OPENCL_FN_clSetKernelArgSVMPointer, cl_int, (cl_kernel p1, cl_uint p2, const void* p3))
 cl_int (CL_API_CALL *clSetKernelArgSVMPointer)(cl_kernel kernel, cl_uint arg_index, const void* arg_value) =
-        opencl_fn3<OPENCL_FN_clSetKernelArgSVMPointer, cl_int, cl_kernel, cl_uint, const void*>::switch_fn;
+        OPENCL_FN_clSetKernelArgSVMPointer_switch_fn;
 static const struct DynamicFnEntry _clSetKernelArgSVMPointer_definition = { "clSetKernelArgSVMPointer", (void**)&clSetKernelArgSVMPointer};
-//void* (CL_API_CALL *clSetKernelExecInfo)(cl_kernel kernel, cl_kernel_exec_info param_name, size_t param_value_size, const void* param_value) =
-//        opencl_fn4<OPENCL_FN_clSetKernelExecInfo, void*, cl_kernel, cl_kernel_exec_info, size_t, const void*>::switch_fn;
-//static const struct DynamicFnEntry _clSetKernelExecInfo_definition = { "clSetKernelExecInfo", (void**)&clSetKernelExecInfo};
-//cl_int (CL_API_CALL *clEnqueueSVMFree)(...) =
-//        opencl_fn8<OPENCL_FN_clEnqueueSVMFree, cl_int, ...>::switch_fn;
-//static const struct DynamicFnEntry _clEnqueueSVMFree_definition = { "clEnqueueSVMFree", (void**)&clEnqueueSVMFree};
+opencl_fn8(OPENCL_FN_clEnqueueSVMMemcpy, cl_int, (cl_command_queue p1, cl_bool p2, void* p3, const void* p4, size_t p5, cl_uint p6, const cl_event* p7, cl_event* p8))
 cl_int (CL_API_CALL *clEnqueueSVMMemcpy)(cl_command_queue command_queue, cl_bool blocking_copy, void* dst_ptr, const void* src_ptr, size_t size, cl_uint num_events_in_wait_list, const cl_event* event_wait_list, cl_event* event) =
-        opencl_fn8<OPENCL_FN_clEnqueueSVMMemcpy, cl_int, cl_command_queue, cl_bool, void*, const void*, size_t, cl_uint, const cl_event*, cl_event*>::switch_fn;
+        OPENCL_FN_clEnqueueSVMMemcpy_switch_fn;
 static const struct DynamicFnEntry _clEnqueueSVMMemcpy_definition = { "clEnqueueSVMMemcpy", (void**)&clEnqueueSVMMemcpy};
+opencl_fn8(OPENCL_FN_clEnqueueSVMMemFill, cl_int, (cl_command_queue p1, void* p2, const void* p3, size_t p4, size_t p5, cl_uint p6, const cl_event* p7, cl_event* p8))
 cl_int (CL_API_CALL *clEnqueueSVMMemFill)(cl_command_queue command_queue, void* svm_ptr, const void* pattern, size_t pattern_size, size_t size, cl_uint num_events_in_wait_list, const cl_event* event_wait_list, cl_event* event) =
-        opencl_fn8<OPENCL_FN_clEnqueueSVMMemFill, cl_int, cl_command_queue, void*, const void*, size_t, size_t, cl_uint, const cl_event*, cl_event*>::switch_fn;
+        OPENCL_FN_clEnqueueSVMMemFill_switch_fn;
 static const struct DynamicFnEntry _clEnqueueSVMMemFill_definition = { "clEnqueueSVMMemFill", (void**)&clEnqueueSVMMemFill};
+opencl_fn8(OPENCL_FN_clEnqueueSVMMap, cl_int, (cl_command_queue p1, cl_bool p2, cl_map_flags p3, void* p4, size_t p5, cl_uint p6, const cl_event* p7, cl_event* p8))
 cl_int (CL_API_CALL *clEnqueueSVMMap)(cl_command_queue command_queue, cl_bool blocking_map, cl_map_flags map_flags, void* svm_ptr, size_t size, cl_uint num_events_in_wait_list, const cl_event* event_wait_list, cl_event* event) =
-        opencl_fn8<OPENCL_FN_clEnqueueSVMMap, cl_int, cl_command_queue, cl_bool, cl_map_flags, void*, size_t, cl_uint, const cl_event*, cl_event*>::switch_fn;
+        OPENCL_FN_clEnqueueSVMMap_switch_fn;
 static const struct DynamicFnEntry _clEnqueueSVMMap_definition = { "clEnqueueSVMMap", (void**)&clEnqueueSVMMap};
+opencl_fn5(OPENCL_FN_clEnqueueSVMUnmap, cl_int, (cl_command_queue p1, void* p2, cl_uint p3, const cl_event* p4, cl_event* p5))
 cl_int (CL_API_CALL *clEnqueueSVMUnmap)(cl_command_queue command_queue, void* svm_ptr, cl_uint num_events_in_wait_list, const cl_event* event_wait_list, cl_event* event) =
-        opencl_fn5<OPENCL_FN_clEnqueueSVMUnmap, cl_int, cl_command_queue, void*, cl_uint, const cl_event*, cl_event*>::switch_fn;
+        OPENCL_FN_clEnqueueSVMUnmap_switch_fn;
 static const struct DynamicFnEntry _clEnqueueSVMUnmap_definition = { "clEnqueueSVMUnmap", (void**)&clEnqueueSVMUnmap};
 
 static const struct DynamicFnEntry* opencl_svm_fn_list[] = {

